lib/vfscore/README.md

```diff
@@ -0,0 +1,108 @@
+# VFSCore: Unikraft's Virtual File System API
+
+`vfscore` is an internal library of Unikraft that provides a generic interface of system calls related to filesystem management.
+There are multiple types of filesystems available in Unikraft (e.g: `ramfs`, `9pfs`).
+
+## Key Functions and Data Structures
+
+The `vfscore` library provides 2 structures to define operations on a filesystem (defined in [`lib/vfscore/include/vfscore/mount.h`](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/include/vfscore/mount.h#L151)):
+
+```c
+struct vfsops {
+        int (*vfs_mount)        (struct mount *, const char *, int, const void*);
+        ...
+        struct vnops    *vfs_vnops;
+};
+```
+
+```c
+struct vnops {
+        vnop_open_t             vop_open;
+        vnop_close_t            vop_close;
+        vnop_read_t             vop_read;
+        vnop_write_t            vop_write;
+        vnop_seek_t             vop_seek;
+        vnop_ioctl_t            vop_ioctl;
+        ...
+};
+```
+
+The first structure mainly defines the operation of mounting the filesystem, while the second defines the operations that can be executed on files (regular files, directories, etc.).
+The `vnops` structure can be seen as the `file_operation` structure in the Linux Kernel (more as an idea).
+More about this structure [here](https://tldp.org/LDP/lkmpg/2.4/html/c577.htm).
+
+The filesystem library will define two such structures through which it will provide the specified operations.
+To understand how these operations end up being used let's examine the [`open` system call](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/syscalls.c#L111)
+
+```c
+int
+sys_open(char *path, int flags, mode_t mode, struct vfscore_file **fpp)
+{
+        struct vfscore_file *fp;
+        struct vnode *vp;
+        ...
+        error = VOP_OPEN(vp, fp);
+}
+```
+
+[`VOP_OPEN()`](https://github.com/unikraft/unikraft/blob/staging/lib/vfscore/include/vfscore/vnode.h#L219) is a macro that is defined as follows:
+
+```c
+#define VOP_OPEN(VP, FP)           ((VP)->v_op->vop_open)(FP)
+```
+
+## Register a new filesystem
+
+To map the desired function for your filesystem with the `vfscore` application, you need to have declared a `struct vnops` [structure](https://github.com/unikraft/unikraft/blob/staging/lib/ramfs/ramfs_vnops.c#L651) and populate its members, as shown below:
```

Comment:
```suggestion
To map the desired function for your filesystem with the `vfscore` application, you need to declare [a `struct vnops` structure](https://github.com/unikraft/unikraft/blob/staging/lib/ramfs/ramfs_vnops.c#L651) and populate its members, as shown below:
```

Reply:
Clarify that `ramfs` is just an example. A filesytem will create its own structure and implement its own functions.